File: social-network/csharp/src/SocialNetwork/Network.cs
namespace SocialNetwork;

public class Network
{
    private readonly IClock _clock;
    private readonly Dictionary<string, User> _users = new();
    private readonly List<Post> _posts = new();

    public Network(IClock clock)
    {
        _clock = clock;
    }

    public IReadOnlyDictionary<string, User> Users => _users;

    public void Post(string userName, string content)
    {
        EnsureRegistered(userName);
        _posts.Add(new Post(userName, content, _clock.Now()));
    }

    public void Follow(string followerName, string followeeName)
    {
        EnsureRegistered(followerName);
        EnsureRegistered(followeeName);
        _users[followerName].Follow(followeeName);
    }

    public IReadOnlyList<Post> Timeline(string userName)
    {
        return _posts
            .Where(p => p.Author == userName)
            .OrderByDescending(p => p.Timestamp)
            .ToList();
    }

    public IReadOnlyList<Post> Wall(string userName)
    {
        var user = _users.GetValueOrDefault(userName);
        if (user == null) return Array.Empty<Post>();

        var visible = new HashSet<string> { userName };
        foreach (var followed in user.Following)
            visible.Add(followed);

        return _posts
            .Where(p => visible.Contains(p.Author))
            .OrderByDescending(p => p.Timestamp)
            .ToList();
    }

    private void EnsureRegistered(string userName)
    {
        if (!_users.ContainsKey(userName))
            _users[userName] = new User(userName);
    }
}

File: social-network/typescript/src/Network.ts
import type { Clock } from './Clock.js';
import { Post } from './Post.js';
import { User } from './User.js';

export class Network {
  private readonly _users = new Map<string, User>();
  private readonly _posts: Post[] = [];

  constructor(private readonly clock: Clock) {}

  get users(): ReadonlyMap<string, User> {
    return this._users;
  }

  post(userName: string, content: string): void {
    this.ensureRegistered(userName);
    this._posts.push(new Post(userName, content, this.clock.now()));
  }

  follow(followerName: string, followeeName: string): void {
    this.ensureRegistered(followerName);
    this.ensureRegistered(followeeName);
    this._users.get(followerName)!.follow(followeeName);
  }

  timeline(userName: string): readonly Post[] {
    return this._posts
      .filter((p) => p.author === userName)
      .sort((a, b) => b.timestamp.getTime() - a.timestamp.getTime());
  }

  wall(userName: string): readonly Post[] {
    const user = this._users.get(userName);
    if (!user) return [];

    const visible = new Set<string>([userName]);
    for (const followed of user.following) {
      visible.add(followed);
    }

    return this._posts
      .filter((p) => visible.has(p.author))
      .sort((a, b) => b.timestamp.getTime() - a.timestamp.getTime());
  }

  private ensureRegistered(userName: string): void {
    if (!this._users.has(userName)) {
      this._users.set(userName, new User(userName));
    }
  }
}
